### ISSUE TITLE
Mix.Xref.calls/0 - list all function calls in an Elixir project

### ISSUE THREAD

**Issue by tmbb** [elixir-lang/elixir#6967]
## High Level Description

`Mix.Tasks.Xref` should provide a public function `Mix.Tasks.Xref.calls/0` that returns data about the function or macro calls in all files in the current project. For each function call, `Mix.Tasks.Xref.calls/0` should return the function that is called (module, function and arity), at least the file where it is called, the module where it is called and line where it is called.

This information could be presented in several ways:

#### As a Flat List (fully normalized)

In this case, the final result would be a flat list of function calls. Each row should be represented as a map, so that adding new keys in future versions doesn't break compatibility. Returning a list of tuples will break compatibility if new keys are added, because pattern matching on tuples is positional. The keys in the map should be:

* `callee` - A tuple containing the module, function and arity of the call
* `line` - Line where the function is called (integer)
* `file` - File where the function is called (binary)
* `caller_module` - Module where the function is called (atom) - this can in principle be deduced from the file, line and information derived from the BEAM debug chunks, but I think it would be convenient to have this information directly accessible here

I leave out of this map information about which function calls our function, because that's pretty easy to deduce accurately from the BEAM debug chunks, but that information could be included too. In that case two other keys could be included:

* `caller_function`
* `caller_arity`

These keys could be `nil` if the function is not called inside another function. Again, these two keys are probably out of the scope of Mix.Xref and should be handled by "userland" code. I include them here mostly for completeness.

This is the fully denormalized representation. It's useful to insert into a database to be queried later, or even "queried" with normal `Enum` functions, like `Enum.filter`. The disadvantage is that it might be inconvenient for some applications.

#### As a Map (denormalized)

Normalizing the data requires choosing an axis over which to aggregate the data. The logical choice is probably the called function. The functions module, name and arity could work as a key for a map with the following structure:

```
%{{m, f, a} => [%{line: line, file: file, caller_module: caller_module}, ...]}
```

This data supports querying only on a single axis ("Where is this function called?"), and would have to be denormalized for other queries. That's why I support the fully denormalized representation. It's preppty simple to turn it into this normalized representation anyway.

## Motivation

The Elixir compiler knows a lot about the interrelations between different parts of the code. Sharing that "knowledge" allows developper tools to be smarter.  Examples of such tools are:

* Source viewers, like my own [Guaxinim](https://github.com/tmbb/guaxinim) (demo here: https://tmbb.github.io/guaxinim/guaxinim.ex.html), which might have hyperlinks from the function calls to their definitions
* Other kinds of visualization tools. For example, there could be graph visualizations at higher level of granularity than what Mix.Xref provides. Those tools cold be implemented by the end users, which would take away work from the core team.
* Refactoring tools, which can tell the developer instantaneously where a certain function is called and possibly rename it.
* Better support of the "jump to definition" functionality in IDEs. Maybe this could be used by the Elixir language server?

## Implementation

This call list can be derived from the `compile_dispatches` and `runtime_dispatches` fields.

Currently, for Elixir 1.5, Guaxinim extracts these data from the compile manifests, so `Mix.Tasks.Xref `can take the same approach. The only overhead this brings is the need to update the `Mix.Tasks.Xref.calls/0` function as the internal representation evolves.

**Comment by josevalim** [comment]
Hi @tmbb! I have edited your proposal to fix some mistakes and slightly change the map format. We should definitely go with a flat list. Any use can do a pass on the data to get the information they want.

I also agree we don't need to include the `caller_function` for now. 

A PR would be appreciated.

**Comment by tmbb** [comment]
> Hi @tmbb! I have edited your proposal to fix some mistakes and slightly change the map format.

Some of the changes in the map are probably for the better. Having the `{m, f, a}` together in a tuple is probably a better fit for how Elixir works, even if it isn't as "denormalized".

On the other hand, I'm not very convinced about the names:

* `call` - a function call is the syntactic construct that means a structure has been called. The call is an interaction with the function, not the function itself. In my mind, a "call" is not identified by a mfa tuple. That tuple identifies the function that is called, not the call. The call is uniquely identified by the mfa tuple + line + file. That's what identifies the call uniquely. You might want to named this [callee](https://en.wiktionary.org/wiki/callee) (from telephony, the one who is called - appropriate given Elixir's heritage). This is the nomenclature used in `Mix.Xref`. If callee is too obscure (I learned the word while reading the mix docs, although the meaning is clearly from the context), at least comething like `called` (but I prefer callee!)

* `caller_line` doesn't sound good. What is it? the line where the function's "caller" lives? What is the function's caller? The function that calls the first function? In that case, shouldn't the `caller_line` de the line where the caller is defined?. I think it should be either `call_line` (the line where the function call - as defined above - exists) or just `line`. The name `caller_line` seems to obfuscate the meaning and is a little ambiguous.

* `caller_file` doesn't sound very good to me (not as bad as above, but still). In my mind, a file is not something that calls. A function may call another function. Even a module may cal a function. But I don't think a file calls anything... The one interpretation that makes sense is "the file where the caller lives". I'd suggest changing this to `call_file`. It's the file where the call lives.

* `caller_module` - I agree with this one. It has two possible meanings: the module that calls the function (I'm ok with modules calling functions) or the module where the "caller" of that function lives.

Again, I agree with the structure of the maps, it's only the names I'm not sure I agree with. What do you think of summoning the "Native English Brigade" here to give their opinion, like you did in the `stream_data` bikeshed discussion?

To summarize, my position regarding the names is:

* `callee` (`{module, function, arity}`) - the function that is called
* `line` or `call_line` - the line where the `callee` is called (where the function call lives)
* `file` or `call_file` - the file where the call lives
* `caller_module` as you've suggested or even `call_module`, the module where the function call lives.

> A PR would be appreciated

I would like to submit such PR, but that requires learning how to run Elixir from source, which might take a while. Otherwise, the code is pretty much written.

**Comment by josevalim** [comment]
Agreed on all naming feedback. I will change your original proposal to keep it in sync, thanks!

> I would like to submit such PR, but that requires learning how to run Elixir from source, which might take a while. Otherwise, the code is pretty much written.

See the README.

**Comment by josevalim** [comment]
An alternative for `callee` is `mfa`. Any preferences?

**Comment by tmbb** [comment]
> An alternative for callee is mfa. Any preferences?

Not really. Both have pros and cons.

* `callee` leaves no confusion on what it is. It's the function that's called, but leaves the actual format obscure.
* `mfa` serves as a mnemonic to remind the reader that it's a 3-tuple with a standard format, but creates confusion on which function it's referring to (the caller? the callee?)

I have a slight bias towards `callee`, but I'm not sure. With `callee`, if in the future we want to add the `caller` we can do so. With `mfa` it would be weirder.

My main preference is that it shouldn't be `callee_mfa` or something like that.

**Comment by AndrewDryga** [comment]
@josevalim What do you think about utilizing this information in a `mix test --stale`? Right now, I guess, we run tests only for a module which is changed, when we could run all tests that would hit changed module and all his dependants.

**Comment by josevalim** [comment]
> Right now, I guess, we run tests only for a module which is changed, when we could run all tests that would hit changed module and all his dependants.

I believe we run all tests that would hit changed modules and all his dependants.

**Comment by josevalim** [comment]
Btw, contributions that add this feature is welcome!

**Comment by tmbb** [comment]
> Btw, contributions that add this feature is welcome!

I haven't forgotten this, but I haven't had much time. I think I can do it in the last week of November.

**Comment by josevalim** [comment]
> I haven't forgotten this, but I haven't had much time. I think I can do it in the last week of November.

Are you OK if someone else does it then?

**Comment by tmbb** [comment]
Sure :)

For anyone interested, it's probably just a matter of integrating these three functions from my [Guaxinim](https://github.com/tmbb/guaxinim) package into Mix.Xref (possibly changing some pattern matches for the new API?):

* https://github.com/tmbb/guaxinim/blob/master/lib/guaxinim/beam_inspector.ex#L75
* https://github.com/tmbb/guaxinim/blob/master/lib/guaxinim/beam_inspector.ex#L66
* https://github.com/tmbb/guaxinim/blob/master/lib/guaxinim/beam_inspector.ex#L45

The functions contain some additional complexity (I use Amnesia to persist the results in a database and some weird columns to serve as lookups, none of which is needed for Mix.Xref to work).

**Comment by jgmchan** [comment]
I can give this a try in the next couple of days, will probably have a bunch of questions soon!

**Comment by jgmchan** [comment]
I've pushed up my current progress to #7026 for this feature. I do have a couple of questions regarding the implementation:
* Compilation functions/macros - Currently we return a list similar to this:
   ```
   [
     %{callee: {A, :a, 0}, caller_module: A, file: "lib/a.ex", line: 2},
     %{callee: {A, :a, 1}, caller_module: A, file: "lib/a.ex", line: 3},
     %{callee: {B, :a, 0}, caller_module: A, file: "lib/a.ex", line: 4},
     %{callee: {Kernel, :def, 2}, caller_module: A, file: "lib/a.ex", line: 4},
     %{callee: {Kernel, :def, 2}, caller_module: A, file: "lib/a.ex", line: 3},
     %{callee: {Kernel, :def, 2}, caller_module: A, file: "lib/a.ex", line: 2},
     %{callee: {Kernel, :defmodule, 2}, caller_module: A, file: "lib/a.ex", line: 1},
     %{callee: {Kernel.LexicalTracker, :read_cache, 2}, caller_module: A, file: "lib/a.ex", line: 1},
     %{callee: {Kernel, :def, 2}, caller_module: B, file: "lib/b.ex", line: 2},
     %{callee: {Kernel, :defmodule, 2}, caller_module: B, file: "lib/b.ex", line: 1},
     %{callee: {Kernel.LexicalTracker, :read_cache, 2}, caller_module: B, file: "lib/b.ex", line: 1}
   ]
   ```
   This contains macros/functions like `Kernel.def/2` and `Kernel.LexicalTracker.read_cache/2` 
   which are called during compilation, but I'm not sure whether they provide any useful information. 
   Is it better to filter these out?

* Umbrella Projects -  Or are we happy to call this function in each umbrella app instead of having it support being called from the parent umbrella project?

**Comment by tmbb** [comment]
> Is it better to filter these out?

No. They provide useful information for projects li Guaxinim or even mix.xref tasks

**Comment by josevalim** [comment]
Agreed with @tmbb.

@jgmchan i don't think we should do anything special for umbrella projects as well besides documenting that it won't work in the umbrella root (since the umbrella root has no manifests).

Other than that, it looks great!

**Comment by jgmchan** [comment]
Ok cool, I'll add a note in the documentation then.

**Comment by josevalim** [comment]
Closing in favor of the PR #7026 .
